examples/src/rescue_raps/custom_trace_table.rs

```diff
@@ -0,0 +1,253 @@
+    ///     into the next row of the trace after the closure returns.
+    pub fn fill<I, U>(&mut self, init: I, update: U)
+    where
+        I: Fn(&mut [B]),
+        U: Fn(usize, &mut [B]),
+    {
+        let mut state = vec![B::ZERO; self.main_trace_width()];
+        init(&mut state);
+        self.update_row(0, &state);
+
+        for i in 0..self.length() - 1 {
+            update(i, &mut state);
+            self.update_row(i + 1, &state);
+        }
+    }
+
+    /// Updates a single row in the execution trace with provided data.
+    pub fn update_row(&mut self, step: usize, state: &[B]) {
+        self.trace.update_row(step, state);
+    }
+
+    // PUBLIC ACCESSORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns the number of columns in this execution trace.
+    pub fn width(&self) -> usize {
+        self.main_trace_width()
+    }
+
+    /// Returns value of the cell in the specified column at the specified row of this trace.
+    pub fn get(&self, column: usize, step: usize) -> B {
+        self.trace.get(column, step)
+    }
+
+    /// Reads a single row from this execution trace into the provided target.
+    pub fn read_row_into(&self, step: usize, target: &mut [B]) {
+        self.trace.read_row_into(step, target);
+    }
+}
+
+// TRACE TRAIT IMPLEMENTATION
+// ================================================================================================
+
+impl<B: StarkField> Trace for RapTraceTable<B> {
+    type BaseField = B;
+
+    fn layout(&self) -> &TraceLayout {
+        &self.layout
+    }
+
+    fn length(&self) -> usize {
+        self.trace.num_rows()
+    }
+
+    fn meta(&self) -> &[u8] {
+        &self.meta
+    }
+
+    fn read_main_frame(&self, row_idx: usize, frame: &mut EvaluationFrame<Self::BaseField>) {
+        let next_row_idx = (row_idx + 1) % self.length();
+        self.trace.read_row_into(row_idx, frame.current_mut());
+        self.trace.read_row_into(next_row_idx, frame.next_mut());
+    }
+
+    fn main_segment(&self) -> &Matrix<B> {
+        &self.trace
+    }
+
+    fn build_aux_segment<E>(
+        &mut self,
+        aux_segments: &[Matrix<E>],
+        rand_elements: &[E],
+    ) -> Option<Matrix<E>>
+    where
+        E: FieldElement<BaseField = Self::BaseField>,
+    {
+        // We only have one auxiliary segment for this example
+        if !aux_segments.is_empty() {
+            return None;
+        }
+
+        let mut row = unsafe { uninit_vector(self.width()) };
+        self.read_row_into(0, &mut row);
+        let mut aux_columns = vec![vec![E::ZERO; self.length()]; self.aux_trace_width()];
+
+        aux_columns[0][0] = rand_elements[0] * row[2].into() + rand_elements[1] * row[3].into();
+        aux_columns[1][0] = rand_elements[0] * row[6].into() + rand_elements[1] * row[7].into();
+
+        // Permutation argument column
+        aux_columns[2][0] = E::ONE;
```

Comment:
This can probably be a single column - but I think having 3 column makes the example simpler to digest. Maybe add a comment that 3 column are technically not required and are used for illustrative purposes?

Reply:
Yes I had kept them for illustrative purposes, comment added!